kvbc/include/categorization/blocks.h

```diff
@@ -0,0 +1,159 @@
+#include "updates.h"
+#include <utility>
+#include <block_digest.h>
+#include "details.h"
+#include "merkle_tree_serialization.h"
+#include "merkle_tree_key_manipulator.h"
+#include "sliver.hpp"
+
+namespace concord::kvbc::categorization {
+
+// Block is composed out of:
+// - BlockID
+// - Parent digest
+// - Categories' updates_info where update info is essentially a list of keys, metadata on the keys and a hash of the
+// category state.
+struct Block {
+  static const std::string CATEGORY_ID;
+
+  Block() = default;
+  Block(const BlockId block_id) { data.block_id = block_id; }
+
+  void add(const std::string& category_id, MerkleUpdatesInfo&& updates_info) {
+    data.categories_updates_info.emplace(std::make_pair(category_id, std::move(updates_info)));
+  }
+
+  void add(const std::string& category_id, KeyValueUpdatesInfo&& updates_info) {
+    data.categories_updates_info.emplace(std::make_pair(category_id, std::move(updates_info)));
+  }
+
+  void add(SharedKeyValueUpdatesInfo&& updates_info) { data.shared_updates_info = std::move(updates_info); }
+
+  void setParentHash(const BlockDigest& parent_hash) {
+    std::copy(parent_hash.begin(), parent_hash.end(), data.parent_digest.begin());
+  }
+  inline BlockId ID() const { return data.block_id; }
+
+  static const serialization::buffer& serialize(const Block& block) {
+    static thread_local serialization::buffer output;
+    output.clear();
+    concord::kvbc::categorization::serialize(output, block.data);
+    return output;
+  }
+
+  static Block deserialize(const serialization::buffer& input) {
+    Block output;
+    concord::kvbc::categorization::deserialize(input, output.data);
+    return output;
+  }
+
+  // Generate block key from block ID
+  // Using CMF for big endian
+  static const serialization::buffer& generateKey(const BlockId block_id) {
+    static thread_local serialization::buffer output;
+    output.clear();
+    BlockKey key{block_id};
+    // think about optimization
+    concord::kvbc::categorization::serialize(output, key);
+    return output;
+  }
+
+  BlockData data;
+};
+
+const std::string Block::CATEGORY_ID{"__blocks__"};
+
+//////////////////////////////////// RAW BLOCKS//////////////////////////////////////
+
+// Reconstructs the updates data as recieved from the user
+// This set methods are overloaded in order to construct the appropriate updates
+
+// Merkle updates reconstruction
+RawBlockMerkleUpdates getRawUpdates(const std::string& category_id,
+                                    const MerkleUpdatesInfo& update_info,
+                                    const BlockId& block_id,
+                                    std::shared_ptr<storage::rocksdb::NativeClient>& native_client) {
```

Comment:
Nit: As there's no ownership and no write ops, we can just pass `const NativeClient*`.

Reply:
yep